### ISSUE TITLE
dotnet-fable RAM usage keep increasing

### ISSUE THREAD

**Issue by MangelMaxime** [fable-compiler/Fable#1087]
### Description

I am running Fable via `dotnet fable start`.

So I never kill the Fable server. When I kill the bundle watcher (rollup in this case) and restart it. I see an increase of the RAM usage of Fable.
I think Fable should release the cause of this memory usage when we kill the bundler.

For example, after one morning of usage I had 1.600 Mo allocated to Fable.

### Related information

* Fable version (`dotnet fable --version`): 1.1.14
* Operating system: Win 10 Pro

**Comment by davidtme** [comment]
It may not be the problem but I’ve hit a memory issue my own f# project. Turns out it was a try catch inside a recursive function. 

https://github.com/fable-compiler/Fable/blob/8c6758be9b0e1d0e8428fa93da62d481dfe94a68/src/dotnet/dotnet-fable/Server.fs#L32

https://stackoverflow.com/questions/13491768/tail-recursion-and-exceptions-in-f

**Comment by alfonsogarciacaro** [comment]
Good point @davidtme, thanks! Unfortunately, [removing try...with from recursive functions](https://github.com/fable-compiler/Fable/commit/f77413523483c0322c96024e7540d8086cafeb52) didn't prevent the memory issue. I had a similar with another .NET app (C#) some time ago. Apparently the app won't run the Garbage Collector until it feels pressure from the OS. If I add [here](https://github.com/fable-compiler/Fable/blob/f77413523483c0322c96024e7540d8086cafeb52/src/dotnet/dotnet-fable/StateUtil.fs#L149) these lines:

```fsharp
        if Map.containsKey project.ProjectFile state then
            GC.Collect()
```

I force garbage collection when a request to compile a project that's already in the state (usually when you stop and restart the bundler again) and the memory usage doesn't go too high. However, it's not a good practice to call `GC.Collect()` by yourself.

If memory usage grows more than 1.5GB for you it means you have a lot of memory available. When the OS sends the app a signal to release memory, you'll see a huge drop in the memory usage. Also, I'm not an expert but when I was analyzing a similar problem in the past I was told that most of that memory (at least in Windows) is being reserved by the app but it's not actually used.

In any case, I'm not really sure about [this part](https://github.com/fable-compiler/Fable/blob/f77413523483c0322c96024e7540d8086cafeb52/src/dotnet/dotnet-fable/StateUtil.fs#L222-L227) where a _pseudo_ thread is spawned to allow parallel compilation. The threads are not killed right after they finish their work (I can see the number of threads adding up for the process in macOS Activity Monitor), so maybe there's a better way (I also tried with `Task` and `Thread`).

@MangelMaxime About removing the project when the bundle is stopped (either because it finishes or because a watch process is cancelled), unfortunately it's not possible because the loader cannot know when the bundle has finished exactly (at least for Webpack). This is the reason why in Fable 1.x the .NET process has to call the Node process and instead of Node calling .NET as it was in Fable 0.7, so Fable can control when the Node process is finished.

> I also tried adding a hook to the `.exit` event in Node, but this event cannot run any async method, so I cannot a TCP message to the Fable daemon to shut it down.
 
In short, projects are kept by Fable state as long as the daemon is running, so if you compile many different projects with the same daemon you can indeed use a lot of memory. However, projects are keyed by the .fproj full path, so when a new request comes to compile the same project again, the memory from the previous one should be freed, but apparently the runtime won't do until it feels the urges from the OS.

**Comment by MangelMaxime** [comment]
Hum ok, that's just I have 95% of my ram used with 25% being fable 😊.

But i agree with what you are saying so perhaps we don't have a solution here.

**Comment by alfonsogarciacaro** [comment]
I'll close this issue for now, we can revisit it if this becomes an actual usability issue :+1:
